### PR TITLE
karavictl list/get powermax

### PR DESCRIPTION
**Pull request by atye**
# Description

This PR adds support for listing/getting PowerMax arrays, filtering the `storage list` output by storage type, and enforcing arguments when getting storage.

**List all storage:**
`karavictl storage list`

**List powerflex storage:**
`karavictl storage list --type powerflex`

**List powermax storage:**
`karavictl storage list --type powermax`

The `karavictl storage get` command now does not default to powerflex. You must supply a storage type and system-id. Otherwise, you will see errors such as `error: system type not specified` and `error: system type not specified`.

There is a new secret in cmd/karavictl/cmd/testdata called `kubectl_get_secret_storage_powerflex_powermax.json` with the following secret data in `storage-systems.yaml`. The system IDs are quoted because otherwise a numeric ID will could be in scientific notation in the JSON output.

```
storage:
  powermax:
    "000197900714":
      Endpoint: https://localhost
      Insecure: true
      Pass: Password123
      User: admin
  powerflex:
    "542a2d5f5122210f":
      Endpoint: https://localhost
      Insecure: true
      Pass: Password123
      User: admin
```

# Issues

List the issues impacted by this PR:

| Issue ID |
| -------- |
|     #72      |

# Checklist:

- [x] I have performed a self-review of my own changes.
```
go test -count=1 -cover -race -timeout 30s -short ./...
?       karavi-authorization/cmd/karavictl      [no test files]
ok      karavi-authorization/cmd/karavictl/cmd  4.301s  coverage: 57.6% of statements
ok      karavi-authorization/cmd/proxy-server   0.067s  coverage: 0.7% of statements
?       karavi-authorization/cmd/sidecar-proxy  [no test files]
?       karavi-authorization/cmd/tenant-service [no test files]
ok      karavi-authorization/deploy     0.095s  coverage: 84.6% of statements
?       karavi-authorization/internal/decision  [no test files]
ok      karavi-authorization/internal/powerflex 9.091s  coverage: 88.9% of statements
ok      karavi-authorization/internal/proxy     6.607s  coverage: 68.0% of statements
ok      karavi-authorization/internal/quota     0.369s  coverage: 92.6% of statements
ok      karavi-authorization/internal/roles     0.048s  coverage: 96.0% of statements
ok      karavi-authorization/internal/tenantsvc 1.805s  coverage: 82.9% of statements
ok      karavi-authorization/internal/token     0.041s  coverage: 90.0% of statements
ok      karavi-authorization/internal/web       0.047s  coverage: 24.1% of statements
?       karavi-authorization/pb [no test files]
```

### CODE REVIEW

**Comment by ioeicicdsvc** [comment]
@atye: Thank you for submitting your pull request!

Please make sure you have followed our contributing guidelines. We will review it as soon as possible.

------------------------------------------------------------------------------------------------------------------------
We want your feedback! If you have any questions or suggestions regarding our contributing process/workflow, please reach out to us at karavi@dell.com.

File: cmd/karavictl/cmd/storage_get_test.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
+package cmd
```
Comment (tdawe) [inline cmd/karavictl/cmd/storage_get_test.go:1]: Add copyright

File: cmd/karavictl/cmd/storage_list_test.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
+package cmd
```
Comment (tdawe) [inline cmd/karavictl/cmd/storage_list_test.go:1]: Add copyright

File: cmd/karavictl/cmd/storage_list.go [hunk 3]
```diff
@@ -63,12 +80,34 @@ var listCmd = &cobra.Command{
...
 	},
 }
 
+func filterStorage(storageType string, allStorage map[string]interface{}) interface{} {
+	if storageType == "" {
+		return allStorage
+	}
+
+	output := make(map[string]interface{})
+	if storage, ok := allStorage["storage"].(map[string]interface{}); ok {
```
Comment (ianbibby) [inline cmd/karavictl/cmd/storage_list.go:97]: Optional: you could reduce most of the indentation of this code by negating the `if`s and utilizing `continue`.

**Review by ianbibby — APPROVED** [review]
👍

**Review by hoppea2 — APPROVED** [review]

File: cmd/karavictl/cmd/storage_get_test.go (hunk 1 above)
Comment (atye) [reply cmd/karavictl/cmd/storage_get_test.go:15]: Fixed.

File: cmd/karavictl/cmd/storage_list_test.go (hunk 2 above)
Comment (atye) [reply cmd/karavictl/cmd/storage_list_test.go:15]: Fixed.

File: cmd/karavictl/cmd/storage_list.go (hunk 3 above)
Comment (atye) [reply cmd/karavictl/cmd/storage_list.go:97]: Refactored.

**Review by hoppea2 — APPROVED** [review]

**Comment by gallacher** [comment]
For consistency, can we quote all that storage system IDs instead of just the PowerMax ones?

File: cmd/karavictl/cmd/testdata/kubectl_get_secret_storage_powerflex_powermax.json [hunk 4]
```diff
@@ -0,0 +1,71 @@
+{
```
Comment (gallacher) [inline cmd/karavictl/cmd/testdata/kubectl_get_secret_storage_powerflex_powermax.json:1]: Is the plan to have all supported storage system test data in this one file? We might want to consider removing the storage system names from the file name if that's the case

Comment (atye) [reply cmd/karavictl/cmd/testdata/kubectl_get_secret_storage_powerflex_powermax.json:1]: There is an existing file called `kubectl_get_secret_storage.json` with just one PowerFlex system. I created this separate file specifically for having a PowerFlex and PowerMax system incase we want to test with various secrets.

**Comment by atye** [comment]
> For consistency, can we quote all that storage system IDs instead of just the PowerMax ones?

We hit this issue because the PowerMax system IDs appear to be all numbers which seem to get interpreted as numbers in the JSON output; thus we see the scientific notation.

If we quote all IDs when creating storage, we'll get escape characters in the JSON output which looks messy. I'll think about this.

```
{
  "storage": {
    "powerflex": {
      "\"09315f941b2a200f\"": {
        "Endpoint": "https://10.247.66.156",
        "Insecure": false,
        "Password": "(omitted)",
        "User": "admin"
      }
    }
  }
}
```

**Review by alikdell — APPROVED** [review]

**Review by ianbibby — APPROVED** [review]

**Comment by atye** [comment]
> > For consistency, can we quote all that storage system IDs instead of just the PowerMax ones?
> 
> We hit this issue because the PowerMax system IDs appear to be all numbers which seem to get interpreted as numbers when we save it in the k8s secret; thus we see the scientific notation when reading it because the IDs start with 000.
> 
> If we quote all IDs when creating storage, we'll get escape characters in the JSON output which looks messy. I'll think about this.
> 
> ```
> {
>   "storage": {
>     "powerflex": {
>       "\"09315f941b2a200f\"": {
>         "Endpoint": "https://10.247.66.156",
>         "Insecure": false,
>         "Password": "(omitted)",
>         "User": "admin"
>       }
>     }
>   }
> }
> ```

This is resolved.

**Review by tdawe — APPROVED** [review]

**Pull request merged**